src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs

```diff
@@ -100,38 +100,173 @@ internal struct RegistrationConfig
+            internal uint ConnFlowControlWindow;
+            internal uint MaxWorkerQueueDelayUs;
+            internal uint MaxStatelessOperations;
+            internal uint InitialWindowPackets;
+            internal uint SendIdleTimeoutMs;
+            internal uint InitialRttMs;
+            internal uint MaxAckDelayMs;
+            internal uint DisconnectTimeoutMs;
+            internal uint KeepAliveIntervalMs;
+            internal ushort PeerBidiStreamCount;
+            internal ushort PeerUnidiStreamCount;
+            internal ushort RetryMemoryLimit;              // Global only
+            internal ushort LoadBalancingMode;             // Global only
+            internal byte MaxOperationsPerDrain;
+            internal QuicSettingsEnabledFlagsFlags EnabledFlags;
+            internal uint* DesiredVersionsList;
+            internal uint DesiredVersionsListLength;
+        }
 
-        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
-        internal delegate void SessionShutdownDelegate(
-            IntPtr session,
-            uint flags,
-            ulong errorCode);
+        [Flags]
+        internal enum QuicSettingsIsSetFlags : ulong
+        {
+            MaxBytesPerKey = 1 << 0,
+            HandshakeIdleTimeoutMs = 1 << 1,
+            IdleTimeoutMs = 1 << 2,
+            TlsClientMaxSendBuffer = 1 << 3,
+            TlsServerMaxSendBuffer = 1 << 4,
+            StreamRecvWindowDefault = 1 << 5,
+            StreamRecvBufferDefault = 1 << 6,
+            ConnFlowControlWindow = 1 << 7,
+            MaxWorkerQueueDelayUs = 1 << 8,
+            MaxStatelessOperations = 1 << 9,
+            InitialWindowPackets = 1 << 10,
+            SendIdleTimeoutMs = 1 << 11,
+            InitialRttMs = 1 << 12,
+            MaxAckDelayMs = 1 << 13,
+            DisconnectTimeoutMs = 1 << 14,
+            KeepAliveIntervalMs = 1 << 15,
+            PeerBidiStreamCount = 1 << 16,
+            PeerUnidiStreamCount = 1 << 17,
+            RetryMemoryLimit = 1 << 18,
+            LoadBalancingMode = 1 << 19,
+            MaxOperationsPerDrain = 1 << 20,
+            SendBufferingEnabled = 1 << 21,
+            PacingEnabled = 1 << 22,
+            MigrationEnabled = 1 << 23,
+            DatagramReceiveEnabled = 1 << 24,
+            ServerResumptionLevel = 1 << 25,
+            DesiredVersionsList = 1 << 26,
+            VersionNegotiationExtEnabled = 1 << 27,
+        }
+
+        [Flags]
+        internal enum QuicSettingsEnabledFlagsFlags : byte
+        {
+            SendBufferingEnabled = 1 << 0,
+            PacingEnabled = 1 << 1,
+            MigrationEnabled = 1 << 2,
+            DatagramReceiveEnabled = 1 << 3,
+            // Contains values of QUIC_SERVER_RESUMPTION_LEVEL
+            ServerResumptionLevel = 1 << 4 | 1 << 5,
+            VersionNegotiationExtEnabled = 1 << 6,
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct CredentialConfig
```

Comment:
This is not right. the "certificate" is union of different structs so whatever is after this will have wrong offset. 
It may not matter right now but the fields after are not safe to use. We should switch the the CredentialConfigCertificateUnion or put in big warning. 

cc @nibanks

Reply:
The certificate is a union of different struct **pointers**, not the objects themselves.

Reply:
I'll not pretend I understand everything in here, but this struct `CredentialConfig` is created to correspond to `QUIC_CREDENTIAL_CONFIG`:
https://github.com/microsoft/msquic/blob/a0dd95474c40ea3f675b6a58083b1149c9b0b19d/src/inc/msquic.h#L251-L264

Underneath is the union `CredentialConfigCertificateUnion` (referenced by `IntPtr` on line 202 here) corresponding to the inner union:
https://github.com/microsoft/msquic/blob/a0dd95474c40ea3f675b6a58083b1149c9b0b19d/src/inc/msquic.h#L254-L260

If I understand it correctly, there are 2 layers: struct with TYPE, FLAGS, etc. and within it a union with different ways to define the certificate itself.

@wfurt could you be more precise with what exactly is wrong and how it should look like?

Note that, we're not actively using any of it ATM.

Reply:
@nibanks is correct. I should stop reading the code in the evening ;(